lib/client/ClientAjax.js

```diff
@@ -0,0 +1,74 @@
+"use strict";
+
+var Promise = require("bluebird");
+var Backbone = require("backbone");
+var BootstrappedDataService = require("./BootstrappedDataService");
+var ClientTestable = require("./ClientTestable");
+var pathJoin = require("../util/pathJoin");
+var _ = require("underscore");
+
+var originalAjax = Backbone.ajax;
+
+var ClientAjax = {
+
+    setup: function(bootstrappedData, appRoot) {
+        ClientTestable.BackboneNativeAjax.Promise = Promise;
+
+        BootstrappedDataService.load(bootstrappedData);
+
+        Backbone.ajax = function(ajaxConfig) {
+            var dataInBootstrappedData = BootstrappedDataService.getFor(ajaxConfig);
+            var bubbledError;
+
+            function bubbleErrorFrom(prop) {
+                var callback = ajaxConfig[prop];
+
+                if (!_.isFunction(callback)) {
+                    return;
+                }
+
+                ajaxConfig[prop] = function() {
+                    try {
```

Comment:
i'm a little concerned too. the workarounds section here https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#2-unsupported-syntax suggests that the deoptimization would be scoped to 30-36 but it's not obvious. performance doesn't seem greatly affected in live testing using chrome timeline so i think it's ok